src/relay/pass/device_annotation.cc

```diff
@@ -0,0 +1,468 @@
+};
+
+/*
+ * \brief Return device allocation map based on the post order traversed graph.
+ * For the following program:
+ * .. code-block:: python
+ *     x = relay.var("x")
+ *     y = relay.var("y")
+ *     add = relay.add(x, y)
+ *     sqrt = relay.sqrt(add)
+ *     log = relay.log(add)
+ *     subtract = relay.subtract(sqrt, log)
+ *     exp = relay.exp(subtract)
+ *
+ * Suppose we have annotated add, sqrt, and log with device 1, 2, and 3,
+ * respectively. The fallback/default device is 4. After Rewriting the
+ * program, we can have the following graph, where each copy op has both
+ * source and destination device id denoting which device the data should be
+ * copied from and to.
+ *
+ *         x     y
+ *          \   /
+ *          add/1
+ *          /   \
+ *       copy1  copy2
+ *         |     |
+ *      sqrt/2 log/3
+ *         |     |
+ *       copy3 copy4
+ *          \   /
+ *        subtract
+ *            |
+ *           exp
+ *
+ * To Get the device mapping of each expression, we need to propagate the
+ * device information from the copy ops. This can be done in two passes.
+ *  -Pass 1: Propagating the source device id to ops in a bottom-up way to the
+ *           ancestors until encountering another copy op. For example, this way
+ *           provides add, x, and y device ids from the copy operator, `copy1`.
+ *  -Pass 2: Propagating the destination device id of "the last" copy op in a
+ *           top-down manner to the nodes on the output paths. For instance,
+ *           this offers `subtract` and `exp` the same device id as `copy3`.
+ */
+
+class DeviceInfo {
+ public:
+  static Map<Expr, Integer> GetDeviceMap(const Expr& expr) {
+    DeviceInfo device_info;
+    device_info.post_visitor_ = PostDfsOrderVisitor();
+    device_info.post_visitor_.Visit(expr);
+    if (device_info.post_visitor_.num_device_copy_ops_ > 0) {
+      device_info.PropagateDeviceId();
+      return device_info.device_map_;
+    } else {
+      return Map<Expr, Integer>();
+    }
+  }
+
+ private:
+  class PostDfsOrderVisitor : private ExprVisitor {
+   public:
+    void Visit(const Expr& expr) { this->VisitExpr(expr); }
+
+   private:
+    // Post order traversal.
+    void VisitExpr_(const FunctionNode* fn) final {
+      ExprVisitor::VisitExpr_(fn);
+      // Skip function node for now.
+      // post_dfs_order_.push_back(fn);
+    }
+
+    void VisitExpr_(const ConstantNode* cn) final {
+      post_dfs_order_.push_back(cn);
+    }
+
+    void VisitExpr_(const CallNode* call) final {
+      // Skip annotation nodes.
+      if (!IsOnDeviceNode(call)) {
+        ExprVisitor::VisitExpr_(call);
+        post_dfs_order_.push_back(call);
+
+        if (IsDeviceCopyNode(call)) {
+          num_device_copy_ops_++;
+        }
+      }
+    }
+
+    void VisitExpr_(const TupleNode* tn) final {
+      ExprVisitor::VisitExpr_(tn);
+      // Skip adding tuple, or keep it but return an array of device ids later.
+      // post_dfs_order_.push_back(tn);
```

Comment:
suggest to print a warning message

Reply:
Logging here seems a little verbose. Will add a TODO for both annotation of function and tuple expressions.